cpu/sam0_common/periph/i2c.c

```diff
@@ -329,48 +422,105 @@ void _i2c_poweroff(i2c_t dev)
 static int _i2c_start(SercomI2cm *dev, uint16_t addr)
 {
     /* Wait for hardware module to sync */
-    DEBUG("Wait for device to be ready\n");
+    DEBUG_PUTS("i2c.c: Wait for device to be ready");
     _syncbusy(dev);
 
     /* Set action to ACK. */
     dev->CTRLB.reg &= ~SERCOM_I2CM_CTRLB_ACKACT;
 
     /* Send Start | Address | Write/Read */
-    DEBUG("Generate start condition by sending address\n");
+    DEBUG("i2c.c: Generate start condition by sending address 0x%04"PRIu16"\n",
+            addr);
     dev->ADDR.reg = addr;
 
-    /* Wait for response on bus. */
-    if (addr & I2C_READ) {
-        /* Some devices (e.g. SHT2x) can hold the bus while
-        preparing the reply */
-        if (_wait_for_response(dev, 100 * SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
-    }
-    else {
-        if (_wait_for_response(dev, SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
-    }
+    /* Wait for response on bus.
+     * Some devices (e.g. SHT2x) can hold the bus while
+     * preparing the reply. */
+    uint32_t timeout = (addr & I2C_READ) ? 100 * SAMD21_I2C_TIMEOUT : SAMD21_I2C_TIMEOUT;
+    int res = _wait_for_response(dev, timeout);
+
+    /* Ignore dev->INTFLAG.bit.ERROR as it does not seem to get set
+     * in the current configuration of the SERCOM module.
+     */
+    bool any_error = dev->STATUS.reg & (
+            SERCOM_I2CM_STATUS_SEXTTOUT /* XXX: do we need this? */ |
+            SERCOM_I2CM_STATUS_ARBLOST | SERCOM_I2CM_STATUS_BUSERR
+            /* not included as always active (hardware bug?) in the
+             * current configuration:
+             * SERCOM_I2CM_STATUS_MEXTTOUT, SERCOM_I2CM_STATUS_LOWTOUT
+             *
+             * not included as only useful with DMA:
+             * SERCOM_I2CM_STATUS_LENERR
+             */
+        ) || (
+            /* bus state should not be unknown at this point */
+            (dev->STATUS.reg & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_UNKNOWN
+        ) || (
+            /* we don't handle multi-master layouts and treat this as error */
+            (dev->STATUS.reg & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_BUSY
+        ) ||
+        (
+            /* slave busy or address not on bus */
+            dev->STATUS.reg & SERCOM_I2CM_STATUS_RXNACK
+        ) || res;
+
+    /* handle errors */
+    if (any_error) {
+        bool unblock_bus = false;
+
+        DEBUG("i2c.c: error: res=%d, INTFLAG=0x%04"PRIu32", STATUS=0x%04"PRIu32"\n",
+            res,
+            (unsigned long)dev->INTFLAG.reg,
+            (unsigned long)dev->STATUS.reg);
+
+        if (res) {
+            DEBUG_PUTS("i2c.c: STATUS_ERR_TIMEOUT");
+            unblock_bus = true;
+        }
+        else if (dev->STATUS.bit.BUSSTATE == BUSSTATE_BUSY) {
```

Comment:
I do like this verbose checking for readability.

I haven't checked, but I'm almost certain the compiler would under normal circumstances optimizes the more verbose code quite well, so that there is no performance penalty compared to less readable and more dense C code. However, these are not normal circumstances because `SercomI2cm::STATUS` is `volatile` and the compiler will be forces to re-load the memory on every access and not combine checks.

How about just reading out the status register into a local variable and check on that? I assume that this would yield optimal or near-optimal machine code. In addition, @dylad will be happy if we cease using the bitfield types (`SERCOM_I2CM_STATUS_Type`), as the CMSIS header files provided by Microchip are no longer generated with bitfield types.

Reply:
I tried to address your comments in my last commit. However, I also read the data sheet (same54/dt51 and saml21) more thoroughly and found a section on address transmission. I reorganized a few of the checks to adhere to that part. The original implementation on master does not completely conform to that part of the data sheet.